src/nodes.coffee

```diff
@@ -2402,79 +2381,108 @@ exports.Assign = class Assign extends Base
             else
-              new PropertyName obj.unwrap().value
-          else
-            # A regular array pattern-match.
-            new Literal expandedIdx or idx
-        name = obj.unwrap().value
-        acc = idx.unwrap() instanceof PropertyName
-        val = new Value new Literal(vvarText), [new (if acc then Access else Index) idx]
-        if defaultValue
-          defaultValue.isDefaultValue = yes
-          val = new Op '?', val, defaultValue
-      if name?
-        message = isUnassignable name
-        obj.error message if message
-      unless obj instanceof Elision
-        assigns.push new Assign(obj, val, null, param: @param, subpattern: yes).compileToFragments o, LEVEL_LIST
+              new PropertyName vvar.unwrap().value
+          acc = idx.unwrap() instanceof PropertyName
+          vval = new Value value, [new (if acc then Access else Index) idx]
+        else
+          # `obj` is [a...], {a...} or a
+          vvar = switch
+            when obj instanceof Splat then new Value obj.name
+            when i in objSpreads then new Value obj.base
+            else obj
+          vval = switch
+            when obj instanceof Splat then compSlice(vvarTxt, i)
+            else new Value new Literal(vvarTxt), [new Index new NumberLiteral i]
+        message = isUnassignable vvar.unwrap().value
+        vvar.error message if message
+        assigns.push new Assign(vvar, vval, null, param: @param, subpattern: yes).compileToFragments o, LEVEL_LIST
+
+    # "Simple" `objects` can be split and compiled to arrays, [a, b, c] = arr, [a, b, c...] = arr
+    assignObjects = (objs, vvarTxt) =>
+      vvar = new Value new Arr(objs, yes)
+      vval = if vvarTxt instanceof Value then vvarTxt else new Value new Literal(vvarTxt)
+      assigns.push new Assign(vvar, vval, null, param: @param, subpattern: yes).compileToFragments o, LEVEL_LIST
+
+    processObjects = (objs, vvarTxt) ->
+      if complexObjects objs
+        loopObjects objs, vvarTxt
       else
-        # Output `Elision` only if `idx` is `i++`, e.g. expandedIdx.
-        assigns.push idx.compileToFragments o, LEVEL_LIST if expandedIdx
-
+        assignObjects objs, vvarTxt
+
+    # In case there is `Splat` or `Expansion` in `objects`,
+    # we can split array in two simple subarrays.
+    # `Splat` [a, b, c..., d, e] can be split into  [a, b, c...] and [d, e].
+    # `Expansion` [a, b, ..., c, d] can be split into [a, b] and [c, d].
+    # Examples:
+    # a) `Splat`
+    #   CS: [a, b, c..., d, e] = arr
+    #   JS: [a, b, ...c] = arr, [d, e] = splice.call(c, -2)
+    # b) `Expansion`
+    #   CS: [a, b, ..., d, e] = arr
+    #   JS: [a, b] = arr, [d, e] = slice.call(arr, -2)
+    if splatsAndExpans.length
+      expIdx = splatsAndExpans[0]
+      leftObjs = objects.slice 0, expIdx + (if isSplat then 1 else 0)
+      rightObjs = objects.slice expIdx + 1
+      processObjects leftObjs, vvarText if leftObjs.length
+      if rightObjs.length
```

Comment:
This is a bit of a pet peeve, but I don’t like `if a.length` as a way of really saying `if a.length isnt 0`. I prefer the latter, because it makes explicit that you’re checking that `length` is nonzero, not that a property name `length` exists. Relying on the fact that when `length` is 0, that 0 evaluates as falsy feels to me like it goes against the “always `===`” principle that is in vogue nowadays and that CoffeeScript enforces with `is` or `==` both equaling `===`.

Reply:
👍